File: Sean的leetcode/python/Count of Range Sum.py
###
# Merge Sort
# Time Complexity: O(nlogn)
# Space Complexity: O(logn) + O(n)
###
class Solution(object):
    def countRangeSum(self, nums, lower, upper):
        """
        :type nums: List[int]
        :type lower: int
        :type upper: int
        :rtype: int
        """
        sums = [0]
        for n in nums:
            sums.append(sums[-1] + n)
        return self.merge_sort(sums, 0, len(sums), lower, upper)
    def merge_sort(self, sums, start, end, lower, upper):
        if end - start <= 1:
            return 0
        mid = (start + end) / 2
        cnt = self.merge_sort(sums, start, mid, lower, upper) + self.merge_sort(sums, mid, end, lower, upper)
        l = mid
        r = mid
        for left in sums[start:mid]:
            while l < end and sums[l] - left < lower:
                l += 1
            while r < end and sums[r] - left <= upper:
                r += 1
            cnt += r - l
        sums[start:end] = sorted(sums[start:end]) #use TimSort to sort the already sorted two subarray, O(n)
        return cnt
                                        
                            
                    

File: MyLeetCode/Leetcode/Count of Range Sum.cpp
//Similar to 315. Count of Smaller Numbers After Self
//Do counting when merge sort
class Solution {
public:
    int countRangeSum(vector<int>& nums, int lower, int upper) {
        int n = nums.size();
        if (n == 0) return 0; 
        vector<long> sums(n+1);
        for (int i = 0; i < n; i++) {
            sums[i+1] = sums[i] + nums[i];
        }
        return sort(sums, 0, n+1, lower, upper);
    }
    int sort(vector<long>& sums, int start, int end, int lower, int upper) {
        if (end - start <= 1) return 0;
        int mid = start + (end - start) / 2;
        int count = sort(sums, start, mid, lower, upper) + sort(sums, mid, end, lower, upper);
        int l = mid, r = mid, t = mid, s = 0;
        
        vector<long> cache(end - start);
        for (int i = start; i < mid; i++) {
            while (l < end && sums[l] - sums[i] < lower) l++; //l, r move from mid to end (increasing)
            while (r < end && sums[r] - sums[i] <= upper) r++;
            count += r - l;
            while (t < end && sums[t] < sums[i]) {
                cache[s++] = sums[t++]; 
            }
            cache[s++] = sums[i];
        }

        for (int i = 0; i < s; i++) {
            sums[start + i] = cache[i];
        }
        return count;
        
        
    }
};